### ISSUE TITLE
[JENKINS-68240] Add BuildNumber as response when a Job is submitted using Jenkins API 

### ISSUE THREAD

**Issue by jenkins-infra-bot** [timja-org/jenkins-gh-issues-poc-11-13#978]
<p>We have a use-case where multiple users can submit requests and our backend needs to process all of them in parallel and continuously check for status of the running job. We did some research on how to associate a job which is submitted with the buildnumber, currently we are using nextBuildnumber concept inorder to get build number and check the status, but this might not work when multiple users are sending requests at the same time. Please let us know if there is any way to achieve this and if I need to make a feature request to get the build number when a job is submitted</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=shravani777">shravani777</a>, imported from: <a href="https://issues.jenkins.io/browse/JENKINS-68240" target="_blank">Add BuildNumber as response when a Job is submitted using Jenkins API </a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Won't Do
<li><b>resolved</b>: 2022-04-23T11:03:29+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-11-13
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>We have a use-case where multiple users can submit requests and our backend needs to process all of them in parallel and continuously check for status of the running job. We did some research on how to associate a job which is submitted with the buildnumber, currently we are using nextBuildnumber concept inorder to get build number and check the status, but this might not work when multiple users are sending requests at the same time. Please let us know if there is any way to achieve this and if I need to make a feature request to get the build number when a job is submitted</p></pre>
</details>
</details>

**Comment by jenkins-infra-bot** [comment]
<sup><i>mwinter69's <a href="https://issues.jenkins.io/browse/JENKINS-68240?focusedId=424319&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-424319">comment</a>:</i></sup>
<p>At the point in time when the job is started the build number is not known. There might be a queue in Jenkins, now imagine user A start the job, then user B and finally user C. Now user A decides to cancel his job waiting in the queue. If you returned the build number at the time it was submitted to the queue the numbers for jobs submitted by B and C would be wrong.</p>

<p>But there is a solution for this when you submit a job you get the ID as part of the response. You can now query the queue for this id</p>


```
<jenkinsurl>/queue/item/<id>/api/xml```
<p>Once the job started you will find in the api information about the execution of the job with the (build)number for a few minutes</p>

<p> </p>

<p> </p>

<pre>

<leftItem _class=<span class="code-quote">"hudson.model.Queue$LeftItem"</span>>
...
<id>10</id>
...
<executable _class=<span class="code-quote">"hudson.model.FreeStyleBuild"</span>>
  <number>1</number>
  <url>http:<span class="code-comment">//localhost:8080/job/demo/1/</url>
</span></executable>
</leftItem>
</pre>
<p>In the api of the job itself you will see the field queueId</p>

<pre>

<freeStyleBuild _class=<span class="code-quote">"hudson.model.FreeStyleBuild"</span>>
...
<queueId>10</queueId>
...
</freeStyleBuild>
</pre>
<p> </p>

<p> </p>
<hr><details><summary>Raw content of original comment</summary>

<pre>
<p>At the point in time when the job is started the build number is not known. There might be a queue in Jenkins, now imagine user A start the job, then user B and finally user C. Now user A decides to cancel his job waiting in the queue. If you returned the build number at the time it was submitted to the queue the numbers for jobs submitted by B and C would be wrong.</p>

<p>But there is a solution for this when you submit a job you get the ID as part of the response. You can now query the queue for this id</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>&lt;jenkinsurl&gt;/queue/item/&lt;id&gt;/api/xml</pre>
</div></div>
<p>Once the job started you will find in the api information about the execution of the job with the (build)number for a few minutes</p>

<p> </p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
&lt;leftItem _class=<span class="code-quote">"hudson.model.Queue$LeftItem"</span>&gt;
...
&lt;id&gt;10&lt;/id&gt;
...
&lt;executable _class=<span class="code-quote">"hudson.model.FreeStyleBuild"</span>&gt;
  &lt;number&gt;1&lt;/number&gt;
  &lt;url&gt;http:<span class="code-comment">//localhost:8080/job/demo/1/&lt;/url&gt;
</span>&lt;/executable&gt;
&lt;/leftItem&gt;
</pre>
</div></div>
<p>In the api of the job itself you will see the field queueId</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
&lt;freeStyleBuild _class=<span class="code-quote">"hudson.model.FreeStyleBuild"</span>&gt;
...
&lt;queueId&gt;10&lt;/queueId&gt;
...
&lt;/freeStyleBuild&gt;
</pre>
</div></div>
<p> </p>

<p> </p></pre>
</details>

<!-- ### Imported Jira references for easier searching -->
<!-- [jira_issue_key=JENKINS-68240] -->
<!-- [jira_comment_id=424319] -->
<!-- [comment_author=mwinter69] -->

**Comment by jenkins-infra-bot** [comment]
<sup><i>mwinter69's <a href="https://issues.jenkins.io/browse/JENKINS-68240?focusedId=424725&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-424725">comment</a>:</i></sup>
<p>Things will stay in the queue for about 5 minutes after the job started afaik, independent of how long the job takes to complete.</p>
<hr><details><summary>Raw content of original comment</summary>

<pre>
<p>Things will stay in the queue for about 5 minutes after the job started afaik, independent of how long the job takes to complete.</p></pre>
</details>
